src/picologging/handlers.py

```diff
@@ -464,3 +464,57 @@ def doRollover(self):
                     addend = 3600
                 now_rollover_at += addend
         self.rollover_at = now_rollover_at
+
+
+class BufferingHandler(picologging.Handler):
+    """
+    A handler class which buffers logging records in memory. Whenever each
+    record is added to the buffer, a check is made to see if the buffer should
+    be flushed. If it should, then flush() is expected to do what's needed.
+    """
+    def __init__(self, capacity):
+        """
+        Initialize the handler with the buffer size.
+        """
+        picologging.Handler.__init__(self)
+        self.capacity = capacity
+        self.buffer = []
+
+    def shouldFlush(self, record):
+        """
+        Should the handler flush its buffer?
+        Returns true if the buffer is up to capacity. This method can be
+        overridden to implement custom flushing strategies.
+        """
+        return (len(self.buffer) >= self.capacity)
+
+    def emit(self, record):
+        """
+        Emit a record.
+        Append the record. If shouldFlush() tells us to, call flush() to process
+        the buffer.
+        """
+        self.buffer.append(record)
+        if self.shouldFlush(record):
+            self.flush()
+
+    def flush(self):
+        """
+        Override to implement custom flushing behaviour.
+        This version just zaps the buffer to empty.
+        """
+        self.acquire()
+        try:
+            self.buffer.clear()
+        finally:
+            self.release()
+
+    def close(self):
+        """
+        Close the handler.
+        This version just flushes and chains to the parent class' close().
+        """
+        try:
+            self.flush()
+        finally:
+            picologging.Handler.close(self)
```

Comment:
I can't remember if Handler.close() actually does anything, if not. just flush

Reply:
https://github.com/python/cpython/blob/cfafd3adf8c3bb73881081f721e77ebfc94431ea/Lib/logging/__init__.py#L1056

> This version removes the handler from an internal map of handlers, _handlers, which is used for handler lookup by name

Right now it won't matter but it would be important if we implement this behaviour for compatibility.

Reply:
ok